File: GPXRouteSimulationJava/app/src/androidTest/java/com/magiclane/sdk/examples/gpxroutesimulationjava/GPXRouteSimulationInstrumentedTests.java
/*
 * SPDX-FileCopyrightText: 2021-2026 Magic Lane International B.V. <info@magiclane.com>
 * SPDX-License-Identifier: Apache-2.0
 *
 * Contact Magic Lane at <info@magiclane.com> for SDK licensing options.
 */

package com.magiclane.sdk.examples.gpxroutesimulationjava;

import android.content.Context;
import android.util.Log;

import androidx.test.core.app.ApplicationProvider;
import androidx.test.filters.LargeTest;

import com.magiclane.sdk.core.GemError;
import com.magiclane.sdk.core.Path;
import com.magiclane.sdk.core.ProgressListener;
import com.magiclane.sdk.examples.testing.GemSdkTestRule;
import com.magiclane.sdk.routesandnavigation.ERouteTransportMode;
import com.magiclane.sdk.routesandnavigation.NavigationListener;
import com.magiclane.sdk.routesandnavigation.NavigationService;
import com.magiclane.sdk.routesandnavigation.RoutingService;
import com.magiclane.sdk.util.GemCall;

import org.junit.Assert;
import org.junit.ClassRule;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.IOException;
import java.io.InputStream;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import androidx.test.internal.runner.junit4.AndroidJUnit4ClassRunner;
import kotlin.Unit;

@LargeTest
@RunWith(AndroidJUnit4ClassRunner.class)
public class GPXRouteSimulationInstrumentedTests {

    private static final Context appContext = ApplicationProvider.getApplicationContext();

    @ClassRule
    public static final GemSdkTestRule sdkRule = new GemSdkTestRule();

    @Test
    public void simulateRoute() throws InterruptedException {
        final String TAG = "GPXSimTest";
        final String[] gpxFiles = {
            "gpx/1.gpx",
            "gpx/2.gpx",
            "gpx/test_route.gpx",
            "gpx/3.gpx",
            "gpx/test_route_old.gpx",
            "gpx/4.gpx",
            "gpx/5.gpx",
            "gpx/test.gpx",
        };

        final NavigationService navigationService = new NavigationService();
        final CountDownLatch latch = new CountDownLatch(gpxFiles.length);
        final RoutingService[] routingServiceHolder = new RoutingService[1];

        final NavigationListener navigationListener = new NavigationListener() {
            @Override
            public void onNavigationInstructionUpdated(
                    com.magiclane.sdk.routesandnavigation.NavigationInstruction instr) {
                GemCall.INSTANCE.execute(() -> {
                    navigationService.cancelNavigation(this);
                    return Unit.INSTANCE;
                });
            }

            @Override
            public void onNavigationError(int error) {
                if (error == GemError.Cancel) {
                    Log.d(TAG, "Navigation cancelled");
                    latch.countDown();
                }
            }
        };

        routingServiceHolder[0] = new RoutingService();
        routingServiceHolder[0].setOnCompleted((routes, errorCode, s) -> {
            if (errorCode == GemError.NoError) {
                com.magiclane.sdk.routesandnavigation.Route route = routes.get(0);
                GemCall.INSTANCE.execute(() -> {
                    ProgressListener progressListener = ProgressListener.Companion.create(
                            null, null, null,
                            code -> {
                                Assert.assertFalse(
                                        GemError.getMessage(code),
                                        GemError.isError(code));
                                return Unit.INSTANCE;
                            },
                            false
                    );
                    int result = navigationService.startSimulationWithRoute(
                            route, navigationListener, progressListener, 1.0f);
                    Assert.assertFalse(GemError.getMessage(result), GemError.isError(result));
                    return Unit.INSTANCE;
                });
            } else {
                Assert.fail(GemError.getMessage(errorCode));
            }
            return Unit.INSTANCE;
        });

        // Trigger first route calculation; each latch.countDown() leads to the next.
        calculateRouteFromGPX(routingServiceHolder[0], gpxFiles[0]);

        // Override latch behaviour: after each navigation cancel, trigger next file.
        // We re-implement a sequential loop using a separate thread.
        Thread sequencer = new Thread(() -> {
            for (int i = 1; i < gpxFiles.length; i++) {
                try {
                    // Wait for previous navigation to finish (cancel received)
                    boolean ok = latch.await(60, TimeUnit.SECONDS);
                    if (!ok) {
                        Assert.fail("Timeout waiting for navigation to cancel for file: " + gpxFiles[i - 1]);
                        return;
                    }
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    return;
                }
                final String gpxFile = gpxFiles[i];
                Log.d(TAG, "Starting: " + gpxFile);
                calculateRouteFromGPX(routingServiceHolder[0], gpxFile);
            }
        });
        sequencer.start();

        // Wait for all navigations to complete (60 seconds total timeout).
        boolean completed = latch.await(60, TimeUnit.SECONDS);
        Assert.assertTrue("Timed out before all simulations finished", completed);
    }

    private static void calculateRouteFromGPX(RoutingService routingService, String gpxAssetPath) {
        GemCall.INSTANCE.execute(() -> {
            try {
                InputStream input = appContext.getResources().getAssets().open(gpxAssetPath);
                Path track = Path.Companion.produceWithGpx(input);
                if (track == null) return Unit.INSTANCE;

                int result = routingService.calculateRoute(
                        track, ERouteTransportMode.Bicycle, false, null, null, null, null);
                Assert.assertFalse(GemError.getMessage(result), GemError.isError(result));
            } catch (IOException e) {
                Assert.fail("IOException opening " + gpxAssetPath + ": " + e.getMessage());
            }
            return Unit.INSTANCE;
        });
    }
}

File: GPXRouteSimulation/app/src/androidTest/kotlin/com/magiclane/sdk/examples/gpxroutesimulation/GPXRouteSimulationInstrumentedTests.kt
// -------------------------------------------------------------------------------------------------------------------------------

/*
 * Copyright (C) 2019-2024, Magic Lane B.V.
 * All rights reserved.
 *
 * This software is confidential and proprietary information of Magic Lane
 * ("Confidential Information"). You shall not disclose such Confidential
 * Information and shall use it only in accordance with the terms of the
 * license agreement you entered into with Magic Lane.
 */

// -------------------------------------------------------------------------------------------------------------------------------

package com.magiclane.sdk.examples.gpxroutesimulation

import android.content.Context
import android.util.Log
import androidx.test.core.app.ApplicationProvider
import androidx.test.filters.LargeTest
import androidx.test.internal.runner.junit4.AndroidJUnit4ClassRunner
import com.magiclane.sdk.core.GemError
import com.magiclane.sdk.core.GemSdk
import com.magiclane.sdk.core.Path
import com.magiclane.sdk.core.ProgressListener
import com.magiclane.sdk.core.SdkSettings
import com.magiclane.sdk.routesandnavigation.ERouteTransportMode
import com.magiclane.sdk.routesandnavigation.NavigationListener
import com.magiclane.sdk.routesandnavigation.NavigationService
import com.magiclane.sdk.routesandnavigation.RoutingService
import com.magiclane.sdk.util.SdkCall
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeout
import org.junit.Assert
import org.junit.BeforeClass
import org.junit.ClassRule
import org.junit.Test
import org.junit.rules.TestRule
import org.junit.runner.Description
import org.junit.runner.RunWith
import org.junit.runners.model.Statement


@LargeTest
@RunWith(AndroidJUnit4ClassRunner::class)
class GPXRouteSimulationInstrumentedTests {
    // -------------------------------------------------------------------------------------------------

    companion object {
        // -------------------------------------------------------------------------------------------------
        const val TIMEOUT = 600000L
        private val appContext: Context = ApplicationProvider.getApplicationContext()
        private var initResult = false

        @get:ClassRule
        @JvmStatic
        val sdkInitRule = SDKInitRule()

        @BeforeClass
        @JvmStatic
        fun checkSdkInitStartActivity() {
            assert(initResult) { "GEM SDK not initialized" }
        }
        // -------------------------------------------------------------------------------------------------
    }

    // -------------------------------------------------------------------------------------------------
    // -------------------------------------------------------------------------------------------------
    class SDKInitRule : TestRule {
        override fun apply(base: Statement, description: Description) = SDKStatement(base)

        inner class SDKStatement(private val base: Statement) : Statement() {
            private val channel = Channel<Boolean>()

            init {
                SdkSettings.onMapDataReady = { isReady ->
                    if (isReady)
                        runBlocking {
                            channel.send(true)
                        }
                }
            }

            @Throws(Throwable::class)
            override fun evaluate() {
                //before tests are executed
                if (!GemSdk.isInitialized()) {
                    runBlocking {
                        initResult = GemSdk.initSdkWithDefaults(appContext)
                        // must wait for map data ready
                        val sdkChannelJob = launch { channel.receive() }
                        withTimeout(TIMEOUT) {
                            while (sdkChannelJob.isActive) delay(500)
                        }
                    }
                } else return

                if (!SdkSettings.isMapDataReady)
                    throw Error(GemError.getMessage(GemError.OperationTimeout))

                try {
                    base.evaluate() // This executes tests
                } finally {
                    GemSdk.release()
                }
            }
        }
    }
    // -------------------------------------------------------------------------------------------------
    // -------------------------------------------------------------------------------------------------

    @Test
    fun simulateRoute(): Unit = runBlocking {
        val channel = Channel<Unit>()
        val navigationService = NavigationService()
        val navigationListener = NavigationListener.create(
            onNavigationInstructionUpdated = {
                SdkCall.execute {
                    navigationService.cancelNavigation()
                }
            },
            onNavigationError = { error ->
                if (error == GemError.Cancel)
                    launch {
                        Log.d("BLABLA", "canceled ")
                        channel.send(Unit)
                    }
            },
            postOnMain = false,
        )
        val routingService = RoutingService(
            onCompleted = { routes, errorCode, _ ->
                when (errorCode) {
                    GemError.NoError -> {
                        val route = routes[0]
                        SdkCall.execute {
                            val result = navigationService.startSimulationWithRoute(
                                route,
                                navigationListener,
                                ProgressListener.create(
                                    onCompleted = { code, _ ->
                                        assert(code == GemError.NoError) { GemError.getMessage(code) }
                                    }
                                )
                            )
                            assert(!GemError.isError(result)) { GemError.getMessage(result) }
                        }
                    }

                    else -> Assert.fail(GemError.getMessage(errorCode))
                }
            }
        )
        launch {
            delay(3000)
            channel.send(Unit)
        }
   /*     launch {
            delay(20000)
        }*/
        withTimeout(60000) {
            val l = arrayListOf(
                "1.gpx",
                "2.gpx",
                "test_route.gpx",
                "3.gpx",
                "test_route_old.gpx",
                "4.gpx",
                "5.gpx",
                "test.gpx",
            )
            l.forEach { gpxAssetPath ->
                channel.receive()
                Log.d("BLABLA", "started $gpxAssetPath")
                calculateRouteFromGPX(routingService, "gpx/$gpxAssetPath")
            }
        }
    }

    private fun calculateRouteFromGPX(routingService: RoutingService, gpxAssetPath: String) =
        SdkCall.execute {
            // Opens GPX input stream.
            val input = appContext.resources.assets.open(gpxAssetPath)

            // Produce a Path based on the data in the buffer.
            val track = Path.produceWithGpx(input/*.readBytes()*/) ?: return@execute

            // Set the transport mode to bike and calculate the route.
            val result = routingService.calculateRoute(track, ERouteTransportMode.Bicycle)
            assert(result == GemError.NoError) { GemError.getMessage(result) }
        }
}
